### PR TITLE
Handle jsx spread attribute with object expression

### PR DESCRIPTION
**Pull request by doniyor2109**
It should solve #539

### CODE REVIEW

File: src/_utils.js [hunk 1]
```diff
@@ -28,22 +28,41 @@ export const addClassName = (path, jsxId) => {
...
+
+        if (~index) {
+          className = attr.get('argument').get(`properties.${index}`)
+          break
+        }
+      }
+
+      if (
+        t.isMemberExpression(node.argument) ||
+        t.isIdentifier(node.argument)
```
Comment (giuseppeg) [inline src/_utils.js]: is this if necessary? what else can we have here? maybe you can add a `continue` inside of `if (t.isObjectExpression(node.argument)) {`

File: test/fixtures/spread-operator.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+const Home = () => (
```
Comment (giuseppeg) [inline test/fixtures/spread-operator.js]: can you remove this fixture file and add a few cases to this https://github.com/zeit/styled-jsx/blob/803896fc36f81d6689ee0f074306fc130a731698/test/fixtures/attribute-generation-classname-rewriting.js ?

```jsx
<div {...{ id: 'foo' }} />
<div {...{ className: 'foo' }} />
<div {...{ className: 'foo' }} className="test" />
<div className="test" {...{ className: 'foo' }} />
<div  {...{ className: 'foo' }} {...bar} />
<div  {...{ className: 'foo' }} {...bar} className="test" />
<div className="test"  {...{ className: 'foo' }} {...bar} />
// more corner cases???
```

File: test/fixtures/attribute-generation-classname-rewriting.js [hunk 3]
```diff
@@ -36,6 +36,10 @@ export default () => {
       <div {...props} {...rest} />
       <div {...props} data-foo {...rest} />
       <div {...props} className={'test'} data-foo {...rest} />
+      <div {...{ className: 'test' }} />
+      <div {...{ className: 'test' }} data-test="test" />
+      <div {...{ className: 'test' }} {...props} data-test="test" />
+      <div {...props} {...{ id: 'test' }} />
```
Comment (giuseppeg) [inline test/fixtures/attribute-generation-classname-rewriting.js]: thanks can you add the other tests form the comment? It is important to make sure that the element gets a `className` prop the spread is merged correctly

Comment (doniyor2109) [reply test/fixtures/attribute-generation-classname-rewriting.js]: My bad, I did not see your tests cases 🙈

Comment (doniyor2109) [reply test/fixtures/attribute-generation-classname-rewriting.js]: I have updated test cases. 
Can you verify that it handles multiple classNames correctly?

File: src/_utils.js (hunk 1 above)
Comment (doniyor2109) [reply src/_utils.js]: > what else can we have here?

You mean other conditions?

Comment (giuseppeg) [reply src/_utils.js]: yea I mean we can spread:

* ObjectExpression
* MemberExpression
* Identifier

so I was thinking that you can add a `continue` inside of the `isObjectExpression` if statement body, this way you don't need this other if statement.

Probably there are other weird edge cases like an inline function call like `{...foo()}` but shrug :P otherwise your code is fine if we want to keep the safety check :) Your call I am fine either ways ;)

File: test/babel6/snapshots/attribute.js.md [hunk 4]
```diff
@@ -146,6 +146,13 @@ Generated by [AVA](https://ava.li).
           <div {...props} {...rest} className={"jsx-2886504620" + " " + (rest.className != null && rest.className || props.className != null && props.className || "")} />␊
           <div {...props} data-foo {...rest} className={"jsx-2886504620" + " " + (rest.className != null && rest.className || props.className != null && props.className || "")} />␊
           <div {...props} data-foo {...rest} className={"jsx-2886504620" + " " + (rest.className != null && rest.className || 'test')} />␊
+          <div {...{ id: 'foo' }} className={"jsx-2886504620"} />␊
+          <div {...{}} className={"jsx-2886504620" + " " + 'foo'} />␊
+          <div {...{ className: 'foo' }} className={"jsx-2886504620" + " " + "test"} />␊
+          <div className="test" {...{}} className={"jsx-2886504620" + " " + 'foo'} />␊
```
Comment (giuseppeg) [inline test/babel6/snapshots/attribute.js.md]: nit: probably in this case we could remove the spread if it is empty

File: test/fixtures/attribute-generation-classname-rewriting.js (hunk 3 above)
Comment (giuseppeg) [reply test/fixtures/attribute-generation-classname-rewriting.js]: looking good! We should add one last test case to make sure that it is handled correctly:

```jsx
<div className="test" {...{ className: props.className }} />
<div className="test" {...{ className: props.className }} {...bar } />
<div className="test" {...bar }  {...{ className: props.className }}  />
```

File: src/_utils.js (hunk 1 above)
Comment (doniyor2109) [reply src/_utils.js]: Understand. If you want I can add function call support too?

File: src/_utils.js [hunk 5]
```diff
@@ -28,12 +28,30 @@ export const addClassName = (path, jsxId) => {
...
+
+        const index = properties.findIndex(
+          property => property.key.name === 'className'
+        )
+
+        if (~index) {
+          className = attr.get('argument').get(`properties.${index}`)
+
+          // Remove jsx spread attribute if there is only className property
+          if (node.argument.properties.length === 1) {
```
Comment (giuseppeg) [inline src/_utils.js]: nit: here you can reference `properties` from L34

File: src/_utils.js [hunk 6]
```diff
@@ -44,9 +62,7 @@ export const addClassName = (path, jsxId) => {
           attrNameDotClassName
         )
       )
-      continue
-    }
-    if (t.isJSXAttribute(attr) && node.name.name === 'className') {
+    } else if (t.isJSXAttribute(attr) && node.name.name === 'className') {
```
Comment (giuseppeg) [inline src/_utils.js]: why this change? I guess you can revert the continue

File: src/_utils.js [hunk 7]
```diff
@@ -28,12 +28,30 @@ export const addClassName = (path, jsxId) => {
...
+        }
+        continue
+      }
+
       const name = node.argument.name
       const attrNameDotClassName = t.memberExpression(
-        t.isMemberExpression(node.argument)
-          ? node.argument
-          : t.identifier(name),
+        t.isIdentifier(node.argument) ? t.identifier(name) : node.argument,
```
Comment (giuseppeg) [inline src/_utils.js]: Sorry I had a second thought, we should probably not handle/support the function call and just let this fail. The reason being that we cannot make assumptions about the function and be sure that it is idempotent i.e. not sure if calling it multiple times  returns always the same thing. I would discourage inline calls like `{...foo()}`

**Comment by giuseppeg** [comment]
@doniyor2109 amazing job! thank you

**Comment by subtleGradient** [comment]
Yay!
